### DISCUSSION TITLE
Script for Benchmarking Exercise

### DISCUSSION THREAD

**Discussion by hkpeaks** [duckdb/duckdb#7609] [Ideas]
I am preparing benchmark for some of high performance software like Polars, Datatable and yours.
I find difficulty to build script similar to the below workflow. I spent a lot of time to learn Polars scripting.
So I want to seek help from each of software developer to provide a script similar to below ETL workflow.
You can find my use cases with runtime for Windows/Linux CLI here https://github.com/hkpeaks/peaks-consolidation/releases

D:\Peaks>do FilterByDifferentCompareOperators100M.txt
Development runtime for testing only
Build Date: 23-05-18 | Expiry Date: 23-08-31
Report Comment: github.com/hkpeaks/peaks-consolidation

Select{100MillionRows.csv | Ledger(=L99,<L20)Project(>B25,<B23)~ Table}
Total Bytes: 7216385229 | Total Batches of Stream: 14
1 2 3 4 5 6 7 8 9 10 11 12 13 14 Table(14 x 9092000)

Select{Currency(!=C06)}
Table(14 x 8096300)

Select{Account(<=11000, >=18000)}
Table(14 x 2477000)

Select{Quantity(Float100..300,Float600..900)}
Table(14 x 1247600)

Select{Contact(C32..C39)}
Table(14 x 929400)

Select{Contact(!=C33)~ Table2}
Table2(14 x 809100)

WriteFile{Table2 ~ PeaksFilterByDifferentCompareOperators100M.csv}
PeaksFilterByDifferentCompareOperators100M.csv(14 x 809100)

Duration: 9.702 seconds

**Comment by l1t1** [comment]
why not use sql?

**Reply by hkpeaks** [reply]
I need to generate results for each step of filter per above. 

And I have try sql statement per below. Not knowing how to produce similar stepwise results similar to above.

@echo off
setlocal EnableDelayedExpansion

set start_time=!time!

:: Your code here
duckdb -c "copy (select * from read_parquet('input/1000MillionRows.parquet') WHERE Ledger>'L50') to 'output/DuckDistinct1000M.parquet';"


set end_time=!time!

echo Start time: %start_time%
echo End time: %end_time%

:: Calculate elapsed time
set /a hours=%end_time:~0,2%-%start_time:~0,2%
set /a minutes=%end_time:~3,2%-%start_time:~3,2%
set /a seconds=%end_time:~6,2%-%start_time:~6,2%

:: Output elapsed time
echo Elapsed time: %hours%:%minutes%:%seconds%

**Comment by l1t1** [comment]
you can modify https://github.com/hkpeaks/peaks-consolidation/blob/main/CompareProgrammingLanguage/Script/PandasReadWriteFile10MillionRows.py
with
https://duckdb.org/docs/archive/0.7.1/api/python/overview

**Comment by l1t1** [comment]
you can modify https://github.com/hkpeaks/peaks-consolidation/blob/main/CompareProgrammingLanguage/Script/PandasReadWriteFile10MillionRows.py
with
https://duckdb.org/docs/archive/0.7.1/api/python/overview

**Comment by hkpeaks** [comment]
Also not knowing how to implement streaming for DuckDB similar to below scripting.

https://github.com/hkpeaks/peaks-consolidation/releases  included some of billion-rows testing results. You can use the runtime to test it. So I want to know if DuckDB doing the same dataset and similar ETL process, the comparsion of performance.

D:\Peaks>do DifferentCommandsStreaming

Select{1000MillionRows.csv | Project(>B28,<B22)Ledger(L10..L12,L53..L55,L82..L85)D/C(=D)~ Table}
Total Bytes: 72163841129 | Total Batches of Stream: 144
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144   Table(14 x 20190000)

Select{Contact(=C39,C32..C34)}
  Table(14 x 7250000)

Select{Quantity(Float500..600)Base Amount(Float>30000)~ FilteredTable}
  FilteredTable(14 x 98000)

ReadFile{Master.csv ~ MasterTable}
  MasterTable(5 x 99213)

BuildKeyValue{MasterTable | Ledger,Account,Project ~ KeyValue}
  KeyValue(3 x 99213)

JoinKeyValue{FilteredTable | Ledger,Account,Project => AllMatch(KeyValue)~ JoinedTable}
  JoinedTable(16 x 98000)

GroupBy{JoinedTable | Cost Centre =>  Count()Max(Quantity)Min(Quantity)Min(Unit Price)Max(Unit Price)Sum(Base Amount)~ Cost Centre}
  Cost Centre(7 x 66)

GroupBy{JoinedTable | Company =>  Count()Max(Exchange Rate)Min(Exchange Rate)Max(Unit Price)Min(Unit Price)Sum(Base Amount)~ Company}
  Company(7 x 34)

WriteFile{FilteredTable ~ PeaksManyCommandsFilterData.csv}
  PeaksManyCommandsFilterData.csv(7 x 98000)

WriteFile{Cost Centre ~ PeaksManyCommandsGroupByCostCentre.csv}
  PeaksManyCommandsGroupByCostCentre.csv(7 x 66)

WriteFile{Company ~ PeaksManyCommandsGroupByCompany.csv}
  PeaksManyCommandsGroupByCompany.csv(7 x 34)

Duration: 124.81 seconds

=============================================================
Above is streaming for ReadFile to Query,

Below is full streaming to cover write file.

PS D:\Peaks> ./do jointablefullstreaming

ReadFile{Master.csv ~ Master}
  Master(5 x 99213)

BuildKeyValue{Master | Ledger,Account,Project ~ KeyValue}
  KeyValue(3 x 99213)

JoinKeyValue{1000MillionRows.csv | Ledger,Account,Project => AllMatch(KeyValue)~ PeaksJoinTableFullStreaming.csv}
Total Bytes: 72163841129 | Total Batches of Stream: 144
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57
 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144   PeaksJoinTableFullStreaming.csv(16 x 1000000000)

Duration: 248.62 seconds
